### PR TITLE
Laboratory work #2, Andrew Zapekin - 23FPL1

### PR DESCRIPTION
**Pull request by Andrew-cmd216**

### CODE REVIEW

File: lab_2_retrieval_w_bm25/main.py [hunk 1]
```diff
@@ -18,7 +17,19 @@ def tokenize(text: str) -> list[str] | None:
 
     In case of corrupt input arguments, None is returned.
     """
+    if not isinstance(text, str):
+        return None
+
+    text2 = ''
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:23]: rename

File: lab_2_retrieval_w_bm25/main.py [hunk 2]
```diff
@@ -18,7 +17,19 @@ def tokenize(text: str) -> list[str] | None:
...
+
+    for i in text.lower():
+        if i.isalpha() or i == ' ':
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:26]: can we remove `i == '  ' ` statement?

File: lab_2_retrieval_w_bm25/main.py [hunk 3]
```diff
@@ -33,7 +44,18 @@ def remove_stopwords(tokens: list[str], stopwords: list[str]) -> list[str] | Non
 
     In case of corrupt input arguments, None is returned.
     """
+    if not isinstance(tokens, list) or not isinstance(stopwords, list) \
+            or not all(isinstance(token, str) for token in tokens) \
+            or not all(isinstance(token, str) for token in stopwords):
+        return None
+
+    lst_without_stops = []
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:52]: may be words_without_sw? or something like that

File: lab_2_retrieval_w_bm25/main.py [hunk 4]
```diff
@@ -33,7 +44,18 @@ def remove_stopwords(tokens: list[str], stopwords: list[str]) -> list[str] | Non
...
+
+    for i in tokens:
+        if i not in stopwords and isinstance(i, str):
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:55]: have you already checked tokens in this list, no?

File: lab_2_retrieval_w_bm25/main.py [hunk 5]
```diff
@@ -33,7 +44,18 @@ def remove_stopwords(tokens: list[str], stopwords: list[str]) -> list[str] | Non
...
     In case of corrupt input arguments, None is returned.
     """
+    if not isinstance(tokens, list) or not isinstance(stopwords, list) \
+            or not all(isinstance(token, str) for token in tokens) \
+            or not all(isinstance(token, str) for token in stopwords):
+        return None
+
+    lst_without_stops = []
+
+    for i in tokens:
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:54]: try to use list comprehension (url: https://www.w3schools.com/python/python_lists_comprehension.asp)

**Review by EIzharikov — COMMENTED** [review]
Andrew, nice start! Сorrect the comments and aim for 10, I wish you good luck!

File: lab_2_retrieval_w_bm25/main.py (hunk 2 above)
Comment (Andrew-cmd216) [reply lab_2_retrieval_w_bm25/main.py:26]: fixed

File: lab_2_retrieval_w_bm25/main.py (hunk 3 above)
Comment (Andrew-cmd216) [reply lab_2_retrieval_w_bm25/main.py:52]: fixed

File: lab_2_retrieval_w_bm25/main.py (hunk 4 above)
Comment (Andrew-cmd216) [reply lab_2_retrieval_w_bm25/main.py:55]: done

File: lab_2_retrieval_w_bm25/main.py (hunk 1 above)
Comment (Andrew-cmd216) [reply lab_2_retrieval_w_bm25/main.py:23]: fixed

File: lab_2_retrieval_w_bm25/main.py (hunk 5 above)
Comment (Andrew-cmd216) [reply lab_2_retrieval_w_bm25/main.py:55]: mb i didn't understand how to use it properly

File: lab_2_retrieval_w_bm25/main.py [hunk 6]
```diff
@@ -47,7 +75,22 @@ def build_vocabulary(documents: list[list[str]]) -> list[str] | None:
 
     In case of corrupt input arguments, None is returned.
     """
+    if not isinstance(documents, list) or documents is None \
+            or not all(isinstance(token, list) for token in documents):
+        return None
+
+    unique_words = []
 
+    for i in documents:
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:84]: rename i

File: lab_2_retrieval_w_bm25/main.py [hunk 7]
```diff
@@ -47,7 +75,22 @@ def build_vocabulary(documents: list[list[str]]) -> list[str] | None:
...
+        if not all(isinstance(token, str) for token in i):
+            return None
+        for j in i:
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:87]: what is j here?? what is i here??

File: lab_2_retrieval_w_bm25/main.py [hunk 8]
```diff
@@ -47,7 +75,22 @@ def build_vocabulary(documents: list[list[str]]) -> list[str] | None:
...
+            if j not in unique_words:
+                unique_words.append(j)
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:89]: can we use sets for unique words?

**Review by EIzharikov — COMMENTED** [review]
Good progress! Make CI green, fix all comments and hurry up! Good luck!

File: lab_2_retrieval_w_bm25/main.py [hunk 9]
```diff
@@ -33,7 +47,21 @@ def remove_stopwords(tokens: list[str], stopwords: list[str]) -> list[str] | Non
...
     """
+    if not isinstance(tokens, list) or not all(isinstance(token, str) for token in tokens):
+        return None
+    if not stopwords or not isinstance(stopwords, list) or \
+            not all(isinstance(stopword, str) for stopword in stopwords):
+        return None
+
+    words_without_sw = []
 
+    for token in tokens:
```
[the remark this answers is not in the document]
Comment (Andrew-cmd216) [reply lab_2_retrieval_w_bm25/main.py:58]: done

File: lab_2_retrieval_w_bm25/main.py (hunk 6 above)
Comment (Andrew-cmd216) [reply lab_2_retrieval_w_bm25/main.py:84]: done

File: lab_2_retrieval_w_bm25/main.py (hunk 7 above)
Comment (Andrew-cmd216) [reply lab_2_retrieval_w_bm25/main.py:87]: fixed

File: lab_2_retrieval_w_bm25/main.py [hunk 10]
```diff
@@ -18,7 +19,20 @@ def tokenize(text: str) -> list[str] | None:
 
     In case of corrupt input arguments, None is returned.
     """
+    if not isinstance(text, str):
+        return None
 
+    words_without_symb = ''
+
+    for i in text.lower():
```
[the remark this answers is not in the document]
Comment (EIzharikov) [reply lab_2_retrieval_w_bm25/main.py:27]: actual

File: lab_2_retrieval_w_bm25/main.py (hunk 8 above)
Comment (EIzharikov) [reply lab_2_retrieval_w_bm25/main.py]: actual

**Comment by EIzharikov** [comment]
You've made some progress, but it's still progress! I wish you good luck in further work!

File: lab_2_retrieval_w_bm25/main.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (EIzharikov) [reply lab_2_retrieval_w_bm25/main.py:27]: ?

File: lab_2_retrieval_w_bm25/main.py [hunk 11]
```diff
@@ -33,7 +47,17 @@ def remove_stopwords(tokens: list[str], stopwords: list[str]) -> list[str] | Non
...
+        return None
+    if not stopwords or not isinstance(stopwords, list) or \
+            not all(isinstance(stopword, str) for stopword in stopwords):
+        return None
+
+    words_without_sw = [token for token in tokens if token not in stopwords]
 
+    if not words_without_sw:
+        return None
+    return words_without_sw
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:60]: try to check for None right in return

File: lab_2_retrieval_w_bm25/main.py [hunk 12]
```diff
@@ -92,7 +172,20 @@ def calculate_tf_idf(tf: dict[str, float], idf: dict[str, float]) -> dict[str, f
...
+
+    tf_idf = {}
+
+    for word in tf:
+        if word in idf:
+            tf_idf[word] = tf[word] * idf[word]
 
+    if not tf_idf:
+        return None
+    return tf_idf
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:188]: try to check for None here too

File: lab_2_retrieval_w_bm25/main.py [hunk 13]
```diff
@@ -92,7 +172,20 @@ def calculate_tf_idf(tf: dict[str, float], idf: dict[str, float]) -> dict[str, f
 
     In case of corrupt input arguments, None is returned.
     """
+    if not tf or not idf or not isinstance(tf, dict) or not isinstance(idf, dict):
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/main.py:175]: join all if checks. The same for other functions

File: lab_2_retrieval_w_bm25/settings.json [hunk 14]
```diff
@@ -1,3 +1,3 @@
 {
-    "target_score": 0
+    "target_score": 8
```
Comment (EIzharikov) [inline lab_2_retrieval_w_bm25/settings.json:2]: Nice aim, maybe 10??

**Review by EIzharikov — COMMENTED** [review]
Fix&answer all my comments, be ready for the next review and make CI green. Good luck!

File: lab_2_retrieval_w_bm25/main.py (hunk 12 above)
Comment (Andrew-cmd216) [reply lab_2_retrieval_w_bm25/main.py:188]: done

File: lab_2_retrieval_w_bm25/main.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (EIzharikov) [reply lab_2_retrieval_w_bm25/main.py:27]: ?

**Review by EIzharikov — COMMENTED** [review]
You have unanswered questions and comments. Fix and answer them, make CI green and be ready for Friday. Good luck

File: lab_2_retrieval_w_bm25/settings.json (hunk 14 above)
Comment (Andrew-cmd216) [reply lab_2_retrieval_w_bm25/settings.json:2]: Don't have enough time for it

File: lab_2_retrieval_w_bm25/main.py (hunk 8 above)
Comment (Andrew-cmd216) [reply lab_2_retrieval_w_bm25/main.py]: done

**Pull request closed without merging**